emote/memory/memory.py

```diff
@@ -463,6 +463,38 @@ def __iter__(self):
             yield {self.data_group: data, self.size_key: data[self.size_key]}
 
 
+class JointMemoryLoader:
+    """A memory loader capable of loading data from multiple `MemoryLoader`s.
+
+    If a datagroup is specified via the `data_group` param, it will place all loaded data in the specified datagroup,
+    otherwise it will directly return the data fetched from each individual loader.
+    """
+
+    def __init__(self, loaders: list[MemoryLoader], data_group: str | None = None):
+        self._loaders = loaders
+        self._data_group = data_group
+
+    def is_ready(self):
+        return all([loader.is_ready() for loader in self._loaders])
+
+    def __iter__(self):
+        if not self.is_ready():
+            raise Exception(
+                """memory loader(s) in JointMemoryLoader does not have enough data. Check `is_ready()`
+                before trying to iterate over data."""
+            )
+
+        while True:
+            out = {}
+            for loader in self._loaders:
+                out.update(next(iter(loader)))
+
+            if self._data_group is not None:
+                out = {self._data_group: out}
```

Comment:
Hmm, it feels like there's some risk of clobbering here... Maybe we can check some trivial errors at least? Ensure the inner data groups don't overlap? Otherwise ensure the keys don't overlap? Or is that something we want? Wondering how we can avoid future hidden errors here.

Reply:
Yeah I think it makes sense to ensure the keys are unique.. Feels like something we only have to do once though, the keys should be static throughout training yes?

Reply:
Ooof. I have no good insights except for saying that the size key has always felt out of place. I don't know what the right solution is and the context switch is too large for me to think hard about emote right now but if we could come up with a nicer structure here that would be really nice. If the interface for memory loaders doesn't work for union memory loaders it feels like we need to think about if we have the right abstraction.

Reply:
I'm wondering whether the size should be part of the group instead of adjacent? I don't remember why we (I?) designed it to be adjacent. Maybe to keep the group pure with data from inside the memory...

Reply:
That sounds fine; can you do a check for which one is used? I think with our decorators for automatic unpacking it'd work fine still, I'm curious if there's some callback that *doesn't* use datagroups that does read the batch size.

Reply:
Adjacent _and_ in sounds like the unclearest of all possible options.

Reply:
I will look into it. Some tests seem to rely on it but probably an easy fix.

Reply:
I think I found why it was placed outside in the first place. The `Trainer` tries to fill the callback state dict with `bp_samples` fetched from a user-defined `batch_size_key`, and it expects a callback to fill in that key:

```
def train(..):
    ...
    for bp_step, batch in zip(counter, self.dataloader):
        ....
        self.state["bp_samples"] += self.state[self._batch_size_key]
```

and that `bp_samples` entry is used for performance logging. If we want to keep this system I think we can solve the clobbering by only injecting a size key on the outermost dataloader. I have updated the joint data loader with this change, wdyt?

Reply:
To me there is something that feels weird about the way we've structured the data here. It feels like the batch size key is misplaced in some way. That said I think your updated code solves all our current problems neatly. Let's go with this and then think more about the problem next time we work on the memory.

Reply:
Do we drive any schedules by bp_samples? If so, they're now much faster. :-)

Reply:
Wait? Why much faster? It seems to me like what Klas did should keep the speed the same?

Reply:
Wouldn't it run by the sum of the batch sizes instead of whichever one was the last one before?

Reply:
I'm not saying it's *wrong*, I was just pointing out that something depending on `bp_samples` progressing at a certain rate the rate would be much higher now.

Reply:
I did not think you were saying that! Yes we should probably address those before merging, I just mean that to me it makes sense to instead rely on bp_step or a particular loaders batch size.

Reply:
I've had a look and could not find anything relying on `bp_samples` so I think this is fine to merge.